srcjs/shinyapp.js

```diff
@@ -714,6 +714,131 @@ var ShinyApp = function() {
     });
   });
 
+  addMessageHandler('shiny-insert-tab', function(message) {
+    var $tabsetPanel = $("#" + message.tabsetPanelId);
+    if ($tabsetPanel.length === 0) {
+      throw 'There is no tabsetPanel with id ' + message.tabsetPanelId;
+    };
+
+    // This is the JS equivalent of the builtItem() R function that is used
```

Comment:
It's unfortunate that we have one copy of this code in R and another copy of it in JS.

I've actually never liked that `tabsetPanel`s modify their contents on the R side. It might make sense to refactor all that code so that the `buildTabset` function is eliminated from the R side, and implemented in JS. To me, that feels like the right place to do the DOM mutations, and it would also allow the code to be reused here.